### PR TITLE
Add support for any TensorSpec-describable object to tf.data.Dataset.from_generator

### PR DESCRIPTION
**Pull request by lithuak**
This PR addresses the #35342 issue which proposes to add the support for different types of tensors to tf.data.Dataset.from_generator. It also finalizes the development effort started at PR #37400.

Initially `from_generator` had two parameters: `output_types` and `output_shapes` (one mandatory and the other one optional) that allowed user to specify the dtypes and optionally shapes of the output objects, but not the types of the objects themselves thus making it impossible for a generator function to output more complicated types of objects like `tf.RaggedTensor` or `tf.SparseTensor`.

We add new parameter `output_signature` which describes the expected output of `from_generator` in terms of [TypeSpecs](https://www.tensorflow.org/api_docs/python/tf/TypeSpec). With this change a user can emit any object from a generator function they decide to pass to the `from_generator` given that such object is describable by the corresponding `TypeSpec` specification.

We leave the `output_types` and `output_shapes` parameters in place for now but mark them as deprecated with the intention to remove them in one of the upcoming TF releases and make `output_signature` the only way to specify the output for `from_generator`.

### CODE REVIEW

**Comment by lithuak** [comment]
@gbaned Hey, could you pls request review from @aaudiber and @edloper ?

File: tensorflow/python/data/ops/dataset_ops.py [hunk 1]
```diff
@@ -831,56 +878,40 @@ def generator_py_func(iterator_id):
...
-        return ret_arrays
-
-      flat_values = script_ops.numpy_function(generator_py_func,
-                                              [iterator_id_t], flattened_types)
-
-      # The `py_func()` op drops the inferred shapes, so we add them back in
-      # here.
-      if output_shapes is not None:
-        for ret_t, shape in zip(flat_values, flattened_shapes):
-          ret_t.set_shape(shape)
```
Comment (edloper) [inline tensorflow/python/data/ops/dataset_ops.py:881]: The old version was calling set_shape on result tensors.  Is the new version doing something equivalent?  (And if so, where?)

**Review by edloper — CHANGES_REQUESTED** [review]

File: tensorflow/python/data/ops/dataset_ops.py [hunk 2]
```diff
@@ -724,28 +724,50 @@ def get_iterator(self, iterator_id):
     def iterator_completed(self, iterator_id):
       del self._iterators[iterator_id]
 
+
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/data/ops/dataset_ops.py]: done

File: tensorflow/python/data/ops/dataset_ops.py [hunk 3]
```diff
@@ -724,28 +724,50 @@ def get_iterator(self, iterator_id):
...
     The `generator` argument must be a callable object that returns
     an object that supports the `iter()` protocol (e.g. a generator function).
-    The elements generated by `generator` must be compatible with the given
-    `output_types` and (optional) `output_shapes` arguments.
 
-    >>> import itertools
-    >>>
+    The elements generated by `generator` must be compatible with either the
+    given `output_signature` argument or with the given `output_types` and
+    (optionally) `output_shapes` arguments whichiver was specified.
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/data/ops/dataset_ops.py]: done

File: tensorflow/python/data/ops/dataset_ops.py [hunk 4]
```diff
@@ -769,31 +791,56 @@ def from_generator(generator, output_types, output_shapes=None, args=None):
...
+        raise TypeError("`output_types` can not be used together with "
+                        "`output_signature`")
+      if output_shapes is not None:
+        raise TypeError("`output_shapes` can not be used together with "
+                        "`output_signature`")
+      if not all(
+          isinstance(_, type_spec.TypeSpec)
+          for _ in nest.flatten(output_signature)):
+        raise TypeError("All the elements of `output_siganture` must be "
+                        "a `tf.TypeSpec` objects.")
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/data/ops/dataset_ops.py]: done

File: tensorflow/python/data/ops/dataset_ops.py [hunk 5]
```diff
@@ -724,28 +724,50 @@ def get_iterator(self, iterator_id):
...
+    The recommended way to call `from_generator` is to use the
+    `output_signature` argument. In this case the output will be assumed to
+    consist of objects with the classes, shapes and types defined by
+    `tf.TypeSpec` objects from `output_signature` argument:
+
     >>> def gen():
-    ...   for i in itertools.count(1):
-    ...     yield (i, [1] * i)
+    ...   ragged_tensor = tf.ragged.constant([[1, 2], [3]],
+    ...                                      ragged_rank=1,
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/data/ops/dataset_ops.py:751]: fixed

File: tensorflow/python/data/kernel_tests/from_generator_test.py [hunk 6]
```diff
@@ -446,6 +453,46 @@ def generator():
...
+    dataset = dataset_ops.Dataset.from_generator(
+        generator,
+        output_signature=ragged_tensor.RaggedTensorSpec(
+            shape=(2, None), dtype=dtypes.int64))
+    get_next = self.getNext(dataset)
+
+    ret = get_next()
+
+    self.assertIsInstance(ret, ragged_tensor.RaggedTensor)
+    self.assertAllEqual([1, 2, 3], ret.values)
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/data/kernel_tests/from_generator_test.py]: fixed

File: tensorflow/python/data/ops/dataset_ops.py [hunk 7]
```diff
@@ -2148,21 +2179,8 @@ def cardinality(self):
...
-    the dataset reads from file(s), the cardinality will be unknown. The
-    cardinality will also be unknown if the dataset contains user-defined
-    functions which could affect the cardinality (such as the functions in
-    `filter`, `flat_map`, `interleave`, or `from_generator`).
-
-    When constructing a dataset, you can apply the
-    `tf.data.experimental.assert_cardinality` transformation to inform the
-    dataset of its expected cardinality, so that `cardinality` can produce a
-    known cardinality.
+    the analysis fails to determine the number of elements in the dataset
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/data/ops/dataset_ops.py:2179]: Nope, it's just some unintentional annoying artifact of series of merges from the previous PR. However, this version corresponds to what we have in the master right now. Don't really understand if/how to fix it.

File: tensorflow/python/ops/script_ops.py [hunk 8]
```diff
@@ -553,7 +571,6 @@ def my_func(x):
     stateful argument making all functions stateful.
     """)
 @tf_export(v1=["py_func"])
-@dispatch.add_dispatch_support
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/ops/script_ops.py]: oh, thank you, that's one very important catch. Incorrect merge again.

File: tensorflow/python/data/util/structure.py [hunk 9]
```diff
@@ -78,24 +78,38 @@ def normalize_element(element):
...
-        spec = type_spec_from_value(t, use_fallback=False)
+        if spec is None:
+          spec = type_spec_from_value(t, use_fallback=False)
+        dtype = getattr(spec, "dtype", None)
       except TypeError:
         # TypeError indicates it was not possible to compute a `TypeSpec` for
         # the value. As a fallback try converting the value to a tensor.
         normalized_components.append(
-            ops.convert_to_tensor(t, name="component_%d" % i))
+            ops.convert_to_tensor(t, name="component_%d" % i, dtype=dtype))
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/data/util/structure.py]: This is it, thanks :) I'd been staring into this trying to figure out why this code looks so wrong to me :)

File: tensorflow/python/data/ops/dataset_ops.py [hunk 10]
```diff
@@ -769,31 +791,56 @@ def from_generator(generator, output_types, output_shapes=None, args=None):
...
+                        "`output_signature`")
+      if not all(
+          isinstance(_, type_spec.TypeSpec)
+          for _ in nest.flatten(output_signature)):
+        raise TypeError("All the elements of `output_siganture` must be "
+                        "a `tf.TypeSpec` objects.")
     else:
-      output_shapes = nest.map_structure_up_to(
-          output_types, tensor_shape.as_shape, output_shapes)
+      if output_types is None and output_shapes is not None:
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/data/ops/dataset_ops.py]: Agreed, fixed.

File: tensorflow/python/data/kernel_tests/from_generator_test.py [hunk 11]
```diff
@@ -259,7 +264,7 @@ def generator():
 
     self.assertAllEqual([1, 2, 3], self.evaluate(get_next()))
     self.assertAllEqual([4, 5, 6], self.evaluate(get_next()))
-    with self.assertRaisesOpError("The expected type was int64"):
+    with self.assertRaises(errors.InvalidArgumentError):
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/data/kernel_tests/from_generator_test.py:267]: This was [changed by request](https://github.com/tensorflow/tensorflow/pull/37400#issuecomment-598952789) of other reviewer. It looks less informative to me too, yet I don't think we should be changing it back. It doesn't look to me that parsing the error string is the right way to communicate the exact error here. It increases the fragility of tests (with immediate example: your proposition to below to slightly change the format of exception string. Could break tests if they relied on parsing the error string). The better way to communicate error here IMHO would be to derive the custom exception class from InvalidArgumentError with something like `expected_signature` and `argument_signature` that hold the incompatible objects or something like that. Could be done as a standalone refactoring PR (could be useful throughout the code in situations like this).

File: tensorflow/python/data/ops/dataset_ops.py [hunk 12]
```diff
@@ -831,56 +878,40 @@ def generator_py_func(iterator_id):
         """A `py_func` that will be called to invoke the iterator."""
         # `next()` raises `StopIteration` when there are no more
         # elements remaining to be generated.
-        values = next(generator_state.get_iterator(iterator_id))
+        values = next(generator_state.get_iterator(iterator_id.numpy()))
+
+        def serialize_structure(s):
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/data/ops/dataset_ops.py]: Thanks, this is so much clearer! Fixed.

File: tensorflow/python/data/ops/dataset_ops.py [hunk 13]
```diff
@@ -831,56 +878,40 @@ def generator_py_func(iterator_id):
...
+                  "expected structure. The expected structure was %s, but the "
+                  "yielded element was %s." %
+                  (serialize_structure(output_signature), values)),
+              sys.exc_info()[2])
+
+        values_spec = structure.type_spec_from_value(values)
+
+        if not structure.are_compatible(values_spec, output_signature):
+          raise TypeError(
+              "`generator` yielded an element of TypeSpec%s where an element "
```
[the remark this answers is not in the document]
Comment (lithuak) [reply tensorflow/python/data/ops/dataset_ops.py]: Fixed.

File: tensorflow/python/data/ops/dataset_ops.py (hunk 1 above)
Comment (lithuak) [reply tensorflow/python/data/ops/dataset_ops.py:881]: The old version use to call `numpy_function()` and exchange numpy ndarrays with it. It used to have the problem of loosing the shapes of them (don't know why), according to the code and the comment. We use `eager_py_func()` instead and exchange Tensors. We don't observe such problem and tests don't show it either.

**Comment by lithuak** [comment]
@edloper Thank you very much for the detailed and thorough review!

**Comment by lithuak** [comment]
@edloper could you please approve or follow up the requested changes?

**Comment by lithuak** [comment]
@aaudiber @jsimsa Hey guys, do you still have plans to release this feature? It's complete, the requested changes are addressed. It's also almost identical to the previous PR that was merged but with the bug fixed and minor changes applied. From the comments it looks like community is still interested in having it.

**Review by edloper — DISMISSED** [review]

**Comment by rthadur** [comment]
@lithuak can you please check sanity builds failures ?

**Comment by lithuak** [comment]
@rthadur Fixed pylint issues, although one of them looks quite mysterious, let's see if it goes away. @edloper Could you try to re-approve please.

**Comment by lithuak** [comment]
@rthadur Fixed pylint issues, although one of them looks quite mysterious, let's see if it goes away. @edloper Could you try to re-approve please.

**Review by rthadur — DISMISSED** [review]

File: tensorflow/python/data/kernel_tests/iterator_test.py [hunk 14]
```diff
@@ -72,7 +72,7 @@ def testCapturingStateInOneShotRaisesException(self):
     dataset = (
         dataset_ops.Dataset.from_tensor_slices([0.0, 1.0, 2.0])
         .map(lambda x: x + var))
-    with self.assertRaisesRegex(
+    with self.assertRaisesRegexp(
```
Comment (jsimsa) [inline tensorflow/python/data/kernel_tests/iterator_test.py:75]: this and similar changes in this file should be reverted

File: tensorflow/python/ops/script_ops.py [hunk 15]
```diff
@@ -80,10 +80,14 @@ def __init__(self, func, Tout, is_grad_func):
         None.
       is_grad_func: Whether this EagerFunc is the gradient of another
         EagerPyFunc.
+      use_tape_cache: (Optional.) Whether to cache `func` in the `tape_cache`.
+        NOTE(lithuak): see the note for `_eager_py_func`.
```
Comment (jsimsa) [inline tensorflow/python/ops/script_ops.py:84]: the `NOTE(...):` is tripping up lint checker. I suggest rewording this to:

For additional information, see description of `_eager_py_func`.

File: tensorflow/python/ops/script_ops.py [hunk 16]
```diff
@@ -370,6 +376,24 @@ def eagerly_executed_grad(*dy):
...
 
+# NOTE(lithuak): this function as a layer of indirection was added with one
+# specific purpose: as a workaround for github issue #35084.
+# It does all the same as `eager_py_func` used to do with one difference:
+# it can be used to instruct underlying EagerFunc not to use `tape_cache`
+# to avoid memory leak. When the issue #35084 is fixed - this function should
+# be removed, its body should be moved back to become the body of
+# `eager_py_func` and all the call sites should be reverted to
+# using `eager_py_func` without `use_tape_cache` argument of any value.
+def _eager_py_func(func, inp, Tout, name=None, use_tape_cache=True):
```
Comment (jsimsa) [inline tensorflow/python/ops/script_ops.py:387]: The lint checker is complaining that this function does not have a docstring. Can you please add a minimal one (which documents the arguments and the purpose of this function). In particular, the "NOTE" above can go be moved to the docstring.

**Comment by gbaned** [comment]
@lithuak  Can you please check @jsimsa's comments and keep us posted ? Thanks!

**Comment by lithuak** [comment]
@jsimsa check it out )

**Review by jsimsa — DISMISSED** [review]

**Comment by rthadur** [comment]
@lithuak  can you please check sanity build failures ?

**Review by rthadur — APPROVED** [review]

**Comment by lithuak** [comment]
@rthadur @jsimsa guys, any clue why copybara fails?

File: tensorflow/python/data/ops/dataset_ops.py [hunk 17]
```diff
@@ -831,56 +875,33 @@ def generator_py_func(iterator_id):
...
         # elements remaining to be generated.
-        values = next(generator_state.get_iterator(iterator_id))
+        values = next(generator_state.get_iterator(iterator_id.numpy()))
 
-        # Use the same _convert function from the py_func() implementation to
-        # convert the returned values to arrays early, so that we can inspect
-        # their values.
         try:
-          flattened_values = nest.flatten_up_to(output_types, values)
+          values = structure.normalize_element(values, output_signature)
```
Comment (aaudiber) [inline tensorflow/python/data/ops/dataset_ops.py:881]: Previously, `from_generator` would convert generated elements to match the types specified with `output_types`, but with this change `from_generator` will require an exact dtype match. The docs say 

> The elements generated by generator must be compatible with the given output_types

https://www.tensorflow.org/api_docs/python/tf/data/Dataset#from_generator

Can we preserve the original behavior of converting compatible dtypes, and add a test to verify that it works?

**Review by aaudiber — CHANGES_REQUESTED** [review]

File: tensorflow/python/data/ops/dataset_ops.py (hunk 17 above)
Comment (lithuak) [reply tensorflow/python/data/ops/dataset_ops.py:881]: @aaudiber Can you give me the example (that we can later convert to the test) of this working in old version and now working in the new one?

Comment (aaudiber) [reply tensorflow/python/data/ops/dataset_ops.py:881]: I looked into it further and I think the current code is still doing implicit conversions between compatible types, so no change is needed. Sorry for the confusion.

**Comment by aaudiber** [comment]
@lithuak, Please update the title and description to directly describe the changes made in the PR. That will help with writing release notes based on the commit history (PR title becomes commit title, and PR body becomes commit description).

File: tensorflow/python/data/ops/dataset_ops.py (hunk 17 above)
Comment (lithuak) [reply tensorflow/python/data/ops/dataset_ops.py:881]: no worries

**Comment by lithuak** [comment]
@aaudiber check it out
